uniffi_bindgen/src/bindings/gecko_js/webidl.rs

```diff
@@ -2,171 +2,537 @@
  * License, v. 2.0. If a copy of the MPL was not distributed with this
  * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
 
-//! Helpers for generating C++ WebIDL bindings from a UniFFI component
-//! interface.
+//! This file contains Gecko JS backend-specific types that wrap the UniFFI
+//! component interface types. These wrappers are used to generate the WebIDL
+//! file and C++ implementation for Firefox.
 //!
-//! The C++ bindings that we generate for Firefox are a little peculiar.
-//! Depending on how they're declared in WebIDL, some methods take extra
-//! arguments. For example, static methods take a `GlobalObject`, methods that
-//! return an `ArrayBuffer` also take a `JSContext`, some return values are
-//! passed via out parameters while others are returned directly, some
-//! arguments map to different C++ types depending on whether they're in or out
-//! parameters, and throwing functions also take an `ErrorResult`.
+//! Although UDL (the UniFFI Definition Language) and WebIDL share the same
+//! syntax, they have completely different semantics. WebIDL distinguishes
+//! between "nullable" (`T?`) and "optional" types (`optional T`), dictionary
+//! members are optional by default, dictionaries are not nullable (but can be
+//! optional, and must have a default value if they are). The UniFFI type
+//! system is much simpler, but can't represent these semantics exactly.
 //!
-//! These conditions and combinations are tricky to express in Askama, so we
-//! handle them in extension traits on the UniFFI types, and keep our templates
-//! clean.
+//!
+//! The Firefox C++ bindings are also peculiar. A C++ function that implements
+//! a WebIDL static or namespace method takes an extra `GlobalObject` argument;
+//! methods that return an `ArrayBuffer` also take a `JSContext`; some return
+//! values are passed via out parameters while others are returned directly;
+//! some WebIDL types map to different C++ types depending on where they're used
+//! (in parameters, out parameters, or dictionary members); and throwing
+//! functions take an extra `ErrorResult` argument.
+//!
+//! https://developer.mozilla.org/en-US/docs/Mozilla/WebIDL_bindings describes
+//! how Firefox WebIDL constructs are reflected in C++. Since UniFFI is
+//! generating both the WebIDL file and the C++ implementation, it must do the
+//! same.
+//!
+//! These combinations of special cases are impossible to express in Askama, so
+//! we have a "shadow type system" that wraps the UniFFI interface types.
+//! Capturing this logic here lets us keep our templates and filters clean.
+
+use crate::interface::{
+    Argument, Constructor, FFIFunction, FFIType, Field, Function, Literal, Method, Object, Record,
+    Type,
+};
+
+/// WebIDL types correspond to UniFFI interface types, s
```

Comment:
trailing `,s` on the end of here, makes we wonder if it's got something important to say :-)

After our conversation earlier, in my head I was lightly expecting to see this implemented as a *trait* rather than as a wrapper type. Is there any interesting context about why a wrapper type makes more sense that is worth capturing in comments here?

(I'm not asking you to try to change it to use traits or anything like that, and "no, there is no interesting context" is a perfectly acceptable answer).

Reply:
Oops, half-complete comment, how embarrassing! 😆

I ran with your idea of using extension traits for `Function`, `Field`, `Argument`, etc.—that actually makes things a lot cleaner, since we don't need to wrap `Object` and `Record` 👍🏻😁  I think that was how I had it originally, and then during some intermediate refactoring (before I added `WebIDLType`) switched to wrappers. I'm not sure `WebIDLType` can be a trait, though—it expands "nullable" to "nullable and optional", so we need to hold an extra variant, and since it's recursive, I think we need to redefine those variants.

I did move `Interface` and `Dictionary` into `Wrapped` (and renamed it to `NonRecursive`), since there's no reason to have those separate. And I renamed `Record` back to `Map`, since that's the UniFFI type name for it.

Thanks for encouraging me to think more about modeling this!